src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -333,24 +336,33 @@ internal MappedField GetOrAddDataField(ImmutableArray<byte> data, ushort alignme
             }
 
             var @this = moduleBuilder.GetPrivateImplClass(syntaxNode, diagnostics);
-            return @this._dataSectionStringLiteralTypes.GetOrAdd(text, static (key, arg) =>
+            return @this._dataSectionStringLiteralTypes.GetOrAdd(text, static (text, arg) =>
             {
-                var (@this, data, diagnostics) = arg;
+                var (@this, data, syntaxNode, diagnostics) = arg;
 
-                string name = "<S>" + DataToHexViaXxHash128(data);
+                string name = "<S>" + @this.DataToHexViaXxHash128(data);
 
                 MappedField dataField = @this.GetOrAddDataField(data, alignment: 1);
 
                 Cci.IMethodDefinition bytesToStringHelper = @this.GetOrSynthesizeBytesToStringHelper(diagnostics);
 
+                var previousText = @this._dataSectionStringLiteralNames.GetOrAdd(name, text);
+                if (previousText != text)
+                {
+                    // If there is a hash collision, we cannot fallback to normal string literal emit strategy
+                    // because the selection of which literal would get which emit strategy would not be deterministic.
```

Comment:
Would it be possible to delay the assignment of the type name that gets emitted into the binary until after the typedef token is assigned? Assuming that typedef tokens are deterministic, the name generated from the typedef token would be deterministic as well. Also, the unique names generated from the typedef tokens can be a lot shorter than the hash.

Reply:
It might be possible, but I'm not sure the compiler's internal object model is ready for that, the Name of the symbol is likely used in other places before the typedef token is assigned.

Chuck has suggested another alternative where we would collect the string literals during binding, sort them by length and content, and then assign indices to them (and names based on that) just before emit.

Also, as I mentioned above, currently we share the machinery for synthesizing data fields with array initializers and u8 literals, and these fields are named using sha256. So ideally we would change those too so they also get names based on indices instead of hashes.

I can add that to the spec as future work and go with an error for now.